addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go

```diff
@@ -827,6 +872,51 @@ func (h *Helper) cloneEmbeddedLocalObjectRef(cluster *clusterapiv1beta1.Cluster,
 	return nil
 }
 
+// AddClusterOwnerRef adds cluster as an owner reference to the children with given controller and blockownerdeletion settings
+func (h *Helper) AddClusterOwnerRef(cluster *clusterapiv1beta1.Cluster, children []*unstructured.Unstructured, controller, blockownerdeletion *bool) error {
+	ownerRef := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               cluster.Kind, // kind is empty after create
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         controller,
+		BlockOwnerDeletion: blockownerdeletion,
+	}
+	return h.setOwnerRef(&ownerRef, children)
+}
+
+func (h *Helper) setOwnerRef(ownerRef *metav1.OwnerReference, children []*unstructured.Unstructured) error {
+	for _, child := range children {
+		gvr, err := h.GVRHelper.GetGVR(child.GroupVersionKind().GroupKind())
+		if err != nil {
+			h.Logger.Error(err, fmt.Sprintf("unable to get GVR of %s/%s", child.GetNamespace(), child.GetName()))
+			return err
+		}
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// We need to get and update, otherwise there could have concurrency issue: ["the object has been modified; please
+			// apply your changes to the latest version and try again"]
+			newChild, errGetProvider := h.DynamicClient.Resource(*gvr).Namespace(child.GetNamespace()).Get(h.Ctx, child.GetName(), metav1.GetOptions{})
+			if errGetProvider != nil {
+				h.Logger.Error(errGetProvider, fmt.Sprintf("unable to get %s %s/%s", child.GetKind(), child.GetNamespace(), child.GetName()))
+				return errGetProvider
+			}
+			newChild = newChild.DeepCopy()
+			newChild.SetOwnerReferences(clusterapiutil.EnsureOwnerRef(newChild.GetOwnerReferences(), *ownerRef))
```

Comment:
Before setting ownerreference can we make sure the provider doesnt have any other ownerreference belonging to a different cluster. If it has throw error. I am thinking instead of webhook we can live with this check for now.

Reply:
added check to see if any of the providers already have the cluster as owner